accepted/0000-retail-pxe-database.md

```diff
@@ -0,0 +1,257 @@
+- Feature Name: Retail info storage
+- Start Date: 2020-12-01
+- RFC PR: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+After deploying Retail (saltboot) client, also known as terminal, PXE entry for this client is generated on Branch server. These entries are not backed by information available on SUSE Manager/Uyuni and are difficult to recreate. This RFC is addressing this concern.
+
+# Motivation
+[motivation]: #motivation
+
+When terminal is booting up, it receives some important information from branch server using kernel command line. In case of first boot, when there is no custom PXE entry for the terminal, these are just information about branch and other custom values. However in case of second and next boot, terminal also receive information about where it should look for salt configuration and also root partition.
+If these information are missing, terminal will consider the boot as first boot and will generate new minion id and salt keys and register as a new client to the server.
+
+
+This is sometimes wanted behaviour, for example when terminal is physically moved from one branch to other branch.
+But this also complicates migration of Uyuni/SUSE Manager to newer versions by installing new branch server as there are more files to backup and restore. If there is some unexpected hardware failure on branch server side, it is difficult to recreate these PXE entries for registered terminals and users are facing a task to reregister deployed terminals.
+
+
+To be able to recreate PXE entries, it is needed to store some additional details about terminals, make them available for the branch server minion and have a state or other mechanism which will recreate them. Idea is that after replacing branch server for any reason, user is able to apply a state or schedule an action and recreate the PXE entries for all connected terminals.
+
+By PXE entries are meant both PXE configuration files commonly named `01-mac_address_dash_separated` and GRUB configuration for UEFI commonly named `01:mac_address_colon_separated.grub2.cfg`.
+
+Later in the text I introduced term `pxe server`. This is usually branch server proxy, however design should be flexible enough to allow any salt client to become pxe server, particularly in branchless scenarios with e.g. peripheral hub servers instead of branch server proxy.
+
+# Detailed design
+[design]: #detailed-design
+
+
+## Workflow description
+
+On each boot (first boot and all next boots) terminal sends an event `suse/manager/pxe_update` together with following data:
```

Comment:
How is this data computed? Some time ago we added the possibility of adding arbitrary grains to the minion start event via:

https://github.com/uyuni-project/uyuni/blob/master/susemanager-utils/susemanager-sls/salt/bootstrap/susemanager.conf#L11

If all this data can be rendered as (custom) grains, then one possibility could be not to use an event at all and just piggyback required data to the minion start event. That would save a round-trip to the client per boot.

Reply:
Startup grains are already used to pass `saltboot_initrd` grain to indicate we are in saltboot initrd mode. However it cannot be used for all pxe data, because terminal will know what boot image to use only after obtaining saltboot and image pillars.